File: app/src/main/java/com/example/jianming/views/YImageView.java
package com.example.jianming.views;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.graphics.Bitmap;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.ImageView;

import java.util.ArrayList;
import java.util.Collection;

public class YImageView extends ImageView {
    private YImageSlider yImageSlider;

    public void setLocationIndex(int locationIndex) {
        this.locationIndex = locationIndex;
    }

    private boolean isDisplay = true;

    public void setDisplay() {
        isDisplay = true;
    }

    public void setNoDisplay() {
        isDisplay = false;
    }

    private int locationIndex;
    private int minX = 0, minY = 0;
    private static final int ANIM_DURATION = 500;

    public YImageView(Context context, YImageSlider yImageSlider, int locationIndex) {
        super(context);
        this.yImageSlider = yImageSlider;
        this.locationIndex = locationIndex;
    }

    int screamH, screamW;

    int originY;

    @Override
    protected boolean setFrame(int l, int t, int r, int b) {
        screamH = b - t;
        screamW = r - l;
        minX = screamW - bitmap_W;
        minY = screamH - bitmap_H;
        if (minY > 0) {
            minY = 0;
        }

        int left, top, right, bottom;
        int contentImageWidth = yImageSlider.getContentView().getBitmap_W();
        if (locationIndex == 1) {
            if (yImageSlider.getAlingLeftOrRight() == 0) {
                left = contentImageWidth + YImageSlider.SPLITE_W;
                top = 0;
                right = contentImageWidth + YImageSlider.SPLITE_W + bitmap_W;
                bottom = bitmap_H;
            } else {
                left = screamW + YImageSlider.SPLITE_W;
                top = 0;
                right = screamW + YImageSlider.SPLITE_W + bitmap_W;
                bottom = bitmap_H;
            }
        } else if (locationIndex == -1) {
            if (yImageSlider.getAlingLeftOrRight() == 0) {
                left = -bitmap_W - YImageSlider.SPLITE_W;
                top = 0;
                right = -YImageSlider.SPLITE_W;
                bottom = bitmap_H;
            } else {
                left = -(bitmap_W + YImageSlider.SPLITE_W + contentImageWidth - screamW);
                top = 0;
                right = -(YImageSlider.SPLITE_W + contentImageWidth - screamW);
                bottom = bitmap_H;
            }
        } else {
            if (yImageSlider.getAlingLeftOrRight() == 0) {
                left = 0;
                top = 0;
                right = bitmap_W;
                bottom = bitmap_H;
            } else {
                left = -(contentImageWidth - screamW);
                top = 0;
                right = screamW;
                bottom = bitmap_H;
            }
        }

        top = (screamH - bitmap_H) / 2;
        originY = top;
        boolean isChanged = super.setFrame(0, 0, bitmap_W, bitmap_H);
        setX(left);
        setY(top);
        return isChanged;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction() & MotionEvent.ACTION_MASK) {

            case MotionEvent.ACTION_DOWN:
                onTouchDown(event);

                break;
            case MotionEvent.ACTION_MOVE:
                onTouchMove(event);
                break;

            case MotionEvent.ACTION_UP:
                onTouchUp();
                break;
        }

        return true;
    }

    AnimatorSet setX, setY, setYE, setXE;


    private void doXAnimationEnd(long duration) {
        if (this.getX() <= 0 && this.getX() >= minX) {
            return;
        }

        float destX;
        if (this.getX() > 0) {
            destX = 0;
        } else {
            destX = minX;
        }
        setXE = new AnimatorSet();
        Collection<Animator> animators = new ArrayList<Animator>();

        animators.add(ObjectAnimator.ofFloat(this, View.X, this.getX(), destX));
        if (yImageSlider.getHideLeft().isDisplay) {
            animators.add(ObjectAnimator.ofFloat(yImageSlider.getHideLeft(), View.X, yImageSlider.getHideLeft().getX(), destX - yImageSlider.getHideLeft().getBitmap_W() - YImageSlider.SPLITE_W));
        }
        if (yImageSlider.getHideRight().isDisplay) {
            animators.add(ObjectAnimator.ofFloat(yImageSlider.getHideRight(), View.X, yImageSlider.getHideRight().getX(), destX + getBitmap_W() + YImageSlider.SPLITE_W));
        }
        setXE.playTogether(animators);
        setXE.setDuration(duration);
        setXE.setInterpolator(new AccelerateInterpolator());
        setXE.start();
    }

    private void doYAnimationEnd(long duration) {
        if (this.getY() <= 0 && this.getY() >= minY) {
            return;
        }

        float destY;

        if (this.getY() > 0) {
            destY = 0;
        } else {
            destY = minY;
        }

        setYE = new AnimatorSet();
//        setYE.play(ObjectAnimator.ofFloat(this, View.Y, this.getY(), destY));
        setYE.playTogether(ObjectAnimator.ofFloat(this, View.Y, this.getY(), destY),
                ObjectAnimator.ofFloat(yImageSlider.getHideRight(), View.Y, yImageSlider.getHideRight().getY(), destY),
                ObjectAnimator.ofFloat(yImageSlider.getHideLeft(), View.Y, yImageSlider.getHideLeft().getY(), destY)
        );
        setYE.setDuration(duration);
        setYE.setInterpolator(new AccelerateInterpolator());
        setYE.start();
    }

    private class AnimData {

        public float dest;
        public boolean useAccelerateInterpolator;
        public long duration;
    }

    private AnimData calAnimDataX(float currPos, int minPos, int velocity) {
        float dest = currPos + velocity * ANIM_DURATION / 2000;
        boolean useAccelerateInterpolator = false;
        if (currPos > 0 || currPos < minPos) {
            useAccelerateInterpolator = true;
            if (currPos > 0) {
                dest = 0;
            } else {
                dest = minPos;
            }
        }

        long aTime;
        long duration = ANIM_DURATION;
        if (dest > 0 || dest < minPos) {
            if (dest > 0) {
                aTime = -(int) currPos * 2 * 1000 / velocity;
            } else {
                aTime = (minPos - (int) currPos) * 2 * 1000 / velocity;
            }
            if (aTime < 0 || aTime > ANIM_DURATION) {
                Log.e("AnimatorSet", "aTime error: " + aTime);
            } else {
                duration = aTime + (ANIM_DURATION - aTime) / 2;
                dest = currPos + velocity * duration / 2000;
            }
        }
        AnimData animData = new AnimData();
        animData.dest = dest;
        animData.useAccelerateInterpolator = useAccelerateInterpolator;
        animData.duration = duration;
        return animData;
    }

    private AnimData calAnimDataY() {
        return calAnimDataY(this.getY(), this.minY, this.velocityY);
    }

    private AnimData calAnimDataY(float currPos, int minPos, int velocity) {
        float dest = currPos + velocity * ANIM_DURATION / 2000;
        boolean useAccelerateInterpolator = false;
        long duration = ANIM_DURATION;
        if (bitmap_H < screamH) {
            dest = originY;
        } else {

            if (currPos > 0 || currPos < minPos) {
                useAccelerateInterpolator = true;
                if (currPos > 0) {
                    dest = 0;
                } else {
                    dest = minPos;
                }
            }

            long aTime;

            if (dest > 0 || dest < minPos) {
                if (dest > 0) {
                    aTime = -(int) currPos * 2 * 1000 / velocity;
                } else {
                    aTime = (minPos - (int) currPos) * 2 * 1000 / velocity;
                }
                if (aTime < 0 || aTime > ANIM_DURATION) {
                    Log.e("AnimatorSet", "aTime error: " + aTime);
                } else {
                    duration = aTime + (ANIM_DURATION - aTime) / 2;
                    dest = currPos + velocity * duration / 2000;
                }
            }
        }
        AnimData animData = new AnimData();
        animData.dest = dest;
        animData.useAccelerateInterpolator = useAccelerateInterpolator;
        animData.duration = duration;
        return animData;
    }

    AnimData animDataX, animDataY;

    public void doBackImgAnim() {
        setX = new AnimatorSet();
        Collection<Animator> animators = new ArrayList<Animator>();

        animators.add(ObjectAnimator.ofFloat(this, View.X, this.getX(), yImageSlider.getHideLeft().getBitmap_W() + YImageSlider.SPLITE_W));
        if (yImageSlider.getHideLeft().isDisplay) {
            animators.add(ObjectAnimator.ofFloat(yImageSlider.getHideLeft(), View.X, yImageSlider.getHideLeft().getX(), 0));
        }
        setX.playTogether(animators);
        setX.setDuration(ANIM_DURATION);
        setX.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                postGetBackImg();
            }
        });
        setX.start();
    }

    public void doNextImgAnim() {
        setX = new AnimatorSet();
        Collection<Animator> animators = new ArrayList<Animator>();

        animators.add(ObjectAnimator.ofFloat(this, View.X, this.getX(), -(this.getBitmap_W() + YImageSlider.SPLITE_W + yImageSlider.getHideRight().getBitmap_W() - screamW)));
        if (yImageSlider.getHideRight().isDisplay) {
            animators.add(ObjectAnimator.ofFloat(yImageSlider.getHideRight(), View.X, yImageSlider.getHideRight().getX(), -(yImageSlider.getHideRight().getBitmap_W() - screamW)));
        }
        setX.playTogether(animators);
        setX.setDuration(ANIM_DURATION);
        setX.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                postGetNextImg();
            }
        });
        setX.start();
    }

    private void onTouchUp() {
        float upX = this.getX();
        if ((upX > screamW / 3 || (upX > 0 && isOnLeftEdge)) && yImageSlider.getHideLeft().isDisplay) {
            doBackImgAnim();
        } else if ((upX + this.getBitmap_W() < screamW * 2 / 3 || (upX < screamW - this.getBitmap_W() && isOnRightEdge)) && yImageSlider.getHideRight().isDisplay) {
            doNextImgAnim();
        } else {
            animDataX = calAnimDataX(this.getX(), minX, velocityX);
            setX = new AnimatorSet();
            Collection<Animator> animators = new ArrayList<Animator>();
            animators.add(ObjectAnimator.ofFloat(this, View.X, this.getX(), animDataX.dest));
            if (yImageSlider.getHideLeft().isDisplay) {
                animators.add(ObjectAnimator.ofFloat(yImageSlider.getHideLeft(), View.X, yImageSlider.getHideLeft().getX(), animDataX.dest - yImageSlider.getHideLeft().getBitmap_W() - YImageSlider.SPLITE_W));
            }
            if (yImageSlider.getHideRight().isDisplay) {
                animators.add(ObjectAnimator.ofFloat(yImageSlider.getHideRight(), View.X, yImageSlider.getHideRight().getX(), animDataX.dest + getBitmap_W() + YImageSlider.SPLITE_W));
            }
            setX.playTogether(animators);
            setX.setDuration(animDataX.duration);

            if (animDataX.useAccelerateInterpolator) {
                setX.setInterpolator(new AccelerateInterpolator());
            } else {
                postXEdgeEvent();
                setX.setInterpolator(new DecelerateInterpolator());
            }

            setX.addListener(new AnimatorListenerAdapter() {
                boolean isCanceled = false;

                @Override
                public void onAnimationEnd(Animator animation) {
                    velocityX = 0;
                    Log.d("", "hideleft.getX()" + yImageSlider.getHideLeft().getX());
                    if (!isCanceled) {
                        if (animDataX.duration < ANIM_DURATION) {
                            postXEdgeEvent();
                            doXAnimationEnd(ANIM_DURATION - animDataX.duration);
                        }
                    }
                }

                @Override
                public void onAnimationCancel(Animator animation) {
                    isCanceled = true;
                }
            });
            setX.start();
        }
        animDataY = calAnimDataY(this.getY(), minY, velocityY);
        AnimData animDataYLeft = yImageSlider.getHideLeft().calAnimDataY();
        AnimData animDataYRight = yImageSlider.getHideRight().calAnimDataY();

        setY = new AnimatorSet();
//        setY.play(ObjectAnimator.ofFloat(this, View.Y, this.getY(), animDataY.dest));
        setY.playTogether(ObjectAnimator.ofFloat(this, View.Y, this.getY(), animDataY.dest),
                ObjectAnimator.ofFloat(yImageSlider.getHideRight(), View.Y, yImageSlider.getHideRight().getY(), animDataYRight.dest),
                ObjectAnimator.ofFloat(yImageSlider.getHideLeft(), View.Y, yImageSlider.getHideLeft().getY(), animDataYLeft.dest)
                );
        setY.setDuration(animDataY.duration);
        if (animDataY.useAccelerateInterpolator) {
            setY.setInterpolator(new AccelerateInterpolator());
        } else {
            postYEdgeEvent();
            setY.setInterpolator(new DecelerateInterpolator());
        }
        setY.addListener(new AnimatorListenerAdapter() {
            boolean isCanceled = false;

            @Override
            public void onAnimationEnd(Animator animation) {
                velocityY = 0;
                if (!isCanceled) {
                    if (animDataY.duration < ANIM_DURATION) {
                        postYEdgeEvent();
                        doYAnimationEnd(ANIM_DURATION - animDataY.duration);
                    }
                }
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                isCanceled = true;
            }
        });
        setY.start();
    }

    private void postGetBackImg() {
        if (edgeListener != null) {
            edgeListener.onGetBackImg(this);
        }
    }

    private void postGetNextImg() {
        if (edgeListener != null) {
            edgeListener.onGetNextImg(this);
        }
    }

    interface EdgeListener {
        void onXEdge(YImageView yImageView);

        void onYEdge(YImageView yImageView);

        void onGetBackImg(YImageView yImageView);

        void onGetNextImg(YImageView yImageView);
    }

    private EdgeListener edgeListener = null;

    public void setEdgeListener(EdgeListener edgeListener) {
        this.edgeListener = edgeListener;
    }

    private void postXEdgeEvent() {
        if (edgeListener != null) {
            edgeListener.onXEdge(this);
        }
    }

    private void postYEdgeEvent() {
        if (edgeListener != null) {
            edgeListener.onYEdge(this);
        }
    }

    boolean isOnLeftEdge = false;
    boolean isOnRightEdge = false;

    int velocityX = 0, velocityY = 0;

    void onTouchDown(MotionEvent event) {
        if (setX != null && setX.isRunning()) {
            setX.cancel();
        }
        if (setY != null && setY.isRunning()) {
            setY.cancel();
        }
        if (setXE != null && setXE.isRunning()) {
            setXE.cancel();
        }
        if (setYE != null && setYE.isRunning()) {
            setYE.cancel();
        }

        isOnLeftEdge = this.getX() >= 0;

        isOnRightEdge = this.getX() <= screamW - getBitmap_W();

        current_x = lastX = event.getRawX();
        current_y = lastY = event.getRawY();
        lastEventTime = event.getEventTime();
    }

    float lastX, lastY;

    long lastEventTime;

    float newX, newY, current_x, current_y;

    void onTouchMove(MotionEvent event) {

        newX = event.getRawX();
        newY = event.getRawY();
        long currEventTime = event.getEventTime();
        if (currEventTime - lastEventTime > 30) {
            float dX = newX - lastX;
            float dY = newY - lastY;
            long dTime = currEventTime - lastEventTime;
            velocityX = (int) (dX * 1000 / dTime);
            velocityY = (int) (dY * 1000 / dTime);
            lastX = newX;
            lastY = newY;
            lastEventTime = currEventTime;
        }

        float currImgX = this.getX();
        float currImgY = this.getY();

        float diffX = (newX - current_x);
        float diffY = (newY - current_y);

        float newImgX = currImgX + diffX;
        float newImgY = currImgY + diffY;
        this.setX(newImgX);
        this.setY(newImgY);

        yImageSlider.getHideLeft().addDiff(diffX, diffY);
        yImageSlider.getHideRight().addDiff(diffX, diffY);


        current_x = (int) event.getRawX();
        current_y = (int) event.getRawY();
    }

    public void addDiff(float diffX, float diffY) {
        setX(getX() + diffX);
        setY(getY() + diffY);
    }

    @Override
    public void setImageBitmap(Bitmap bm) {
        super.setImageBitmap(bm);
        bitmap_W = bm.getWidth();
        bitmap_H = bm.getHeight();
    }

    public int getBitmap_W() {
        return bitmap_W;
    }

    private int bitmap_W, bitmap_H;

}

File: app/src/main/java/org/nanking/knightingal/kslideviewlib/YImageView.kt
package org.nanking.knightingal.kslideviewlib

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import androidx.appcompat.widget.AppCompatImageView

import java.util.ArrayList

@SuppressLint("ViewConstructor")
class YImageView(
    context: Context,
    private val yImageSlider: YImageSlider,
    var locationIndex: Int,
    id: Int
) : AppCompatImageView(context) {
    companion object {
        private const val TAG = "YImageView"
        private const val ANIM_DURATION:Long = 500
    }

    var isDisplay = true
    var vx = 0.toFloat()
    var vy = 0.toFloat()

    private var lastX = 0.toFloat()
    private var lastY = 0.toFloat()
    private var lastEventTime = 0.toLong()
    private var currentX = 0.toFloat()
    private var currentY = 0.toFloat()
    private var newX = 0.toFloat()
    private var newY = 0.toFloat()
    private var bitmapW:Int = 0
    private var bitmapH:Int = 0


    private var minX = 0
    private var minY = 0

    private var screamH:Int = 0
    private var screamW:Int = 0

    private var originY:Int = 0

    private val fitAllImageSize: Boolean = true

    private var setX: AnimatorSet = AnimatorSet()
    private var setY: AnimatorSet = AnimatorSet()
    private var setYE: AnimatorSet = AnimatorSet()
    private var setXE: AnimatorSet = AnimatorSet()
    lateinit var animDataX:AnimData
    lateinit var animDataY:AnimData

    lateinit var postGetBackImg: ()->Unit
    lateinit var postGetNextImg: ()->Unit

    lateinit var postXEdgeEvent: ()->Unit
    lateinit var postYEdgeEvent: ()->Unit

    private var isOnLeftEdge = false
    private var isOnRightEdge = false

    override fun setFrame(l:Int, t:Int, r:Int, b:Int):Boolean {
        screamH = b - t
        screamW = r - l
        minX = screamW - bitmapW
        minY = if (screamH - bitmapH > 0) 0 else screamH - bitmapH

        if (bitmapW < screamW || fitAllImageSize) {
            val rat = (bitmapH.toFloat())/(bitmapW.toFloat())
            bitmapW = screamW
            bitmapH = (rat * bitmapW).toInt()
        }

        val contentImageWidth = yImageSlider.contentView.bitmapW

        val left = calLeftPoint(contentImageWidth)

        val right = calRightPoint(contentImageWidth)

        val top = (screamH - bitmapH) / 2
        originY = top
        val bottom = top + bitmapH
        Log.d(TAG, "left=$left, right=$right, top=$top, bottom=$bottom")

        val isChanged = super.setFrame(0, top, bitmapW, bottom)
        x = left.toFloat()
        y = top.toFloat()
        return isChanged;
    }

    private fun calLeftPoint(contentImageWidth: Int): Int {
        when (locationIndex) {
            1 -> {
                return if (yImageSlider.alingLeftOrRight == 0) {
                    contentImageWidth + YImageSlider.SPLITE_W
                } else {
                    screamW + YImageSlider.SPLITE_W
                }
            }
            -1 -> {
                return if (yImageSlider.alingLeftOrRight == 0) {
                    -bitmapW - YImageSlider.SPLITE_W
                } else {
                    -(bitmapW + YImageSlider.SPLITE_W + contentImageWidth - screamW)
                }
            }
            else -> {
                return if (yImageSlider.alingLeftOrRight == 0) {
                    0
                } else {
                    -(contentImageWidth - screamW)
                }
            }
        }
    }

    private fun calRightPoint(contentImageWidth: Int): Int {
        when (locationIndex) {
            1 -> {
                return if (yImageSlider.alingLeftOrRight == 0) {
                    contentImageWidth + YImageSlider.SPLITE_W + bitmapW
                } else {
                    screamW + YImageSlider.SPLITE_W + bitmapW
                }
            }
            -1 -> {
                return if (yImageSlider.alingLeftOrRight == 0) {
                    -YImageSlider.SPLITE_W
                } else {
                    -(YImageSlider.SPLITE_W + contentImageWidth - screamW)
                }
            }
            else -> {
                return if (yImageSlider.alingLeftOrRight == 0) {
                    bitmapW
                } else {
                    screamW
                }
            }
        }
    }

    override fun onTouchEvent(event:MotionEvent):Boolean {
        when (event.action.and(MotionEvent.ACTION_MASK)) {
            MotionEvent.ACTION_DOWN -> onTouchDown(event)
            MotionEvent.ACTION_MOVE -> onTouchMove(event)
            MotionEvent.ACTION_UP -> onTouchUp()
        }
        return true
    }



    private fun  doXAnimationEnd(duration: Long) {
        if (x.toInt() in minX..0) {
            return
        }

        val destX:Float = if (x > 0) 0.toFloat() else minX.toFloat()

        setXE = AnimatorSet()

        val animators = ArrayList<Animator>()
        animators.add(ObjectAnimator.ofFloat(this, View.X, x, destX))
        animators.add(ObjectAnimator.ofFloat(
                yImageSlider.hideLeft,
                View.X,
                yImageSlider.hideLeft.x,
                destX - yImageSlider.hideLeft.bitmapW - YImageSlider.SPLITE_W
        ))
        animators.add(ObjectAnimator.ofFloat(
                yImageSlider.hideRight,
                View.X,
                yImageSlider.hideRight.x,
                destX + bitmapW + YImageSlider.SPLITE_W
        ))

        setXE.playTogether(animators)
        setXE.duration = duration
        setXE.interpolator = AccelerateInterpolator()
        setXE.start()
    }

    private fun  doYAnimationEnd(duration: Long) {
        if (y.toInt() in minY..0) {
            return
        }
        val destY = if (y > 0) 0 else minY

        setYE = AnimatorSet()
        setYE.playTogether(ObjectAnimator.ofFloat(this, View.Y, y, destY.toFloat()))
        setYE.duration = duration
        setYE.interpolator = AccelerateInterpolator()
        setYE.start()

    }

    private fun hideImageAnimBack() {
        val setYE = AnimatorSet()
        setYE.playTogether(
                ObjectAnimator.ofFloat(
                        yImageSlider.hideRight,
                        View.Y,
                        yImageSlider.hideRight.y,
                        yImageSlider.hideRight.originY.toFloat()
                ),
                ObjectAnimator.ofFloat(
                        yImageSlider.hideLeft,
                        View.Y,
                        yImageSlider.hideLeft.y,
                        yImageSlider.hideLeft.originY.toFloat()
                )
        )
        setYE.duration = ANIM_DURATION
        setYE.interpolator = AccelerateInterpolator()
        setYE.start()
    }

    data class AnimData(var dest:Float, var useAccelerateInterpolator:Boolean, var duration:Long)

    private fun  calAnimDataX(currPos: Float, minPos: Int, velocity: Float): YImageView.AnimData {
        var dest = currPos + velocity * ANIM_DURATION / 2000
        var useAccelerateInterpolator = false
        if (currPos > 0 || currPos < minPos) {
            useAccelerateInterpolator = true
            dest = if (currPos > 0) {
                0.toFloat()
            } else {
                minPos.toFloat()
            }
        }


        var duration = ANIM_DURATION
        if (dest > 0 || dest < minPos) {
            val aTime = if (dest > 0)
                (-currPos * 2 * 1000 / velocity).toLong()
            else
                ((minPos.toFloat() - currPos) * 2 * 1000 / velocity).toLong()

            if (aTime < 0 || aTime > ANIM_DURATION) {
                Log.e(TAG, "aTime error: $aTime")
            } else {
                duration = aTime + (ANIM_DURATION - aTime) / 2
                dest = currPos + velocity * duration / 2000
            }
        }
        return AnimData(dest, useAccelerateInterpolator, duration)
    }


    private fun  calAnimDataY(currPos: Float, minPos: Int, velocity: Float): YImageView.AnimData {
        var dest = currPos + velocity * ANIM_DURATION / 2000
        var useAccelerateInterpolator = false
        var duration = ANIM_DURATION
        if (bitmapH < screamH) {
            dest = originY.toFloat()
        } else {

            if (currPos > 0 || currPos < minPos) {
                useAccelerateInterpolator = true
                dest = if (currPos > 0) {
                    0.toFloat()
                } else {
                    minPos.toFloat()
                }
            }

            if (dest > 0 || dest < minPos) {
                val aTime = if (dest > 0)
                    (-currPos * 2 * 1000 / velocity).toLong()
                else
                    ((minPos.toFloat() - currPos) * 2 * 1000 / velocity).toLong()
                if (aTime < 0 || aTime > ANIM_DURATION) {
                    Log.e(TAG, "aTime error: $aTime")
                } else {
                    duration = aTime + (ANIM_DURATION - aTime) / 2
                    dest = currPos + velocity * duration / 2000
                }
            }
        }
        return AnimData(dest, useAccelerateInterpolator, duration)
    }


    fun doBackImgAnim() {
        setX = AnimatorSet()
        val animators = ArrayList<Animator>()

        animators.add(ObjectAnimator.ofFloat(
                this,
                View.X,
                x,
                (yImageSlider.hideLeft.bitmapW + YImageSlider.SPLITE_W).toFloat()
        ))
        animators.add(ObjectAnimator.ofFloat(
                this,
                View.Y,
                y,
                originY.toFloat()
        ))
        Log.d(TAG, "doBackImgAnim, originY=$originY")
        animators.add(ObjectAnimator.ofFloat(
                yImageSlider.hideLeft, View.X, yImageSlider.hideLeft.x, 0.toFloat()
        ))
        animators.add(ObjectAnimator.ofFloat(
                yImageSlider.hideLeft, View.Y, yImageSlider.hideLeft.y, yImageSlider.hideLeft.originY.toFloat()
        ))
        setX.playTogether(animators)
        setX.duration = ANIM_DURATION
        setX.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                postGetBackImg()
            }
        })
        setX.start()
    }

    fun doNextImgAnim() {
        setX = AnimatorSet()
        val animators = ArrayList<Animator>()

        animators.add(ObjectAnimator.ofFloat(
                this,
                View.X,
                x,
                -(bitmapW
                        + YImageSlider.SPLITE_W
                        + yImageSlider.hideRight.bitmapW
                        - screamW).toFloat()
        ))
        Log.d(TAG, "doNextImgAnim, originY=$originY")
        animators.add(ObjectAnimator.ofFloat(
                this,
                View.Y,
                y,
                originY.toFloat()
        ))
        animators.add(ObjectAnimator.ofFloat(
                yImageSlider.hideRight,
                View.X,
                yImageSlider.hideRight.x,
                -(yImageSlider.hideRight.bitmapW - screamW).toFloat()
        ))
        animators.add(ObjectAnimator.ofFloat(
                yImageSlider.hideRight,
                View.Y,
                yImageSlider.hideRight.y,
                yImageSlider.hideRight.originY.toFloat()
        ))
        setX.playTogether(animators)
        setX.duration = ANIM_DURATION
        setX.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                postGetNextImg()
            }
        })
        setX.start()
    }

    private fun onTouchUp() {
        val upx = x
        if (upx > 0 && yImageSlider.hideLeft.isDisplay) {
            doBackImgAnim()
        } else if (upx < screamW - this.bitmapW && yImageSlider.hideRight.isDisplay) {
            doNextImgAnim()
        } else if (upx > screamW / 3 && yImageSlider.hideLeft.isDisplay) {
            doBackImgAnim()
        } else if (upx + this.bitmapW < screamW * 2 / 3 && yImageSlider.hideRight.isDisplay) {
            doNextImgAnim()
        } else {
            animDataX = calAnimDataX(x, minX, vx)
            setX = AnimatorSet()
            val animators = ArrayList<Animator>()
            animators.add(ObjectAnimator.ofFloat(this, View.X, x, animDataX.dest))
            animators.add(ObjectAnimator.ofFloat(
                    yImageSlider.hideLeft,
                    View.X,
                    yImageSlider.hideLeft.x,
                    animDataX.dest - yImageSlider.hideLeft.bitmapW - YImageSlider.SPLITE_W
            ))
            animators.add(ObjectAnimator.ofFloat(
                    yImageSlider.hideRight,
                    View.X,
                    yImageSlider.hideRight.x,
                    animDataX.dest + bitmapW + YImageSlider.SPLITE_W
            ))
            setX.playTogether(animators)
            setX.duration = animDataX.duration
            if (animDataX.useAccelerateInterpolator) {
                setX.interpolator = AccelerateInterpolator()
            } else {
                postXEdgeEvent()
                setX.interpolator = DecelerateInterpolator()
            }

            setX.addListener(object : AnimatorListenerAdapter() {
                var isCanceled = false

                override fun onAnimationEnd(animation: Animator) {
                    vx = 0.toFloat()
                    if (!isCanceled) {
                        if (animDataX.duration < ANIM_DURATION) {
                            postXEdgeEvent()
                            doXAnimationEnd(ANIM_DURATION - animDataX.duration)
                        }
                    }

                }

                override fun onAnimationCancel(animation: Animator) {
                    isCanceled = true
                }
            })
            setX.start()
        }
        animDataY = calAnimDataY(y, minY, vy)

        setY = AnimatorSet()
        setY.playTogether(ObjectAnimator.ofFloat(
                this,
                View.Y,
                y,
                animDataY.dest
        ))
        setY.duration = animDataY.duration
        setY.interpolator = if (animDataY.useAccelerateInterpolator)
            AccelerateInterpolator()
        else
            DecelerateInterpolator()
        if (!animDataY.useAccelerateInterpolator) {
            postYEdgeEvent()
        }
        setY.addListener(object :AnimatorListenerAdapter() {
            var isCanceled = false

            override fun onAnimationEnd(animation: Animator) {
                vy = 0.toFloat()
                if (!isCanceled) {
                    if (animDataY.duration < ANIM_DURATION) {
                        postYEdgeEvent()
                        doYAnimationEnd(ANIM_DURATION - animDataY.duration)
                    }
                }
            }

            override fun onAnimationCancel(animation: Animator) {
                isCanceled = true
            }
        })
        setY.start()
        hideImageAnimBack()
    }



    private fun  onTouchDown(event: MotionEvent) {
        if (setX.isRunning ) {
            setX.cancel()
        }
        if (setY.isRunning) {
            setY.cancel()
        }
        if (setXE.isRunning) {
            setXE.cancel()
        }
        if (setYE.isRunning) {
            setYE.cancel()
        }

        isOnLeftEdge = this.x >= 0

        isOnRightEdge = this.y <= screamW - bitmapW

        currentX = event.rawX
        lastX = event.rawX

        currentY = event.rawY
        lastY = event.rawY
        lastEventTime = event.eventTime
    }


    private fun onTouchMove(event: MotionEvent) {
        newX = event.rawX
        newY = event.rawY
        val currentEventTime = event.eventTime
        if (currentEventTime - lastEventTime > 30) {
            val dx = newX - lastX
            val dy = newY - lastY
            val dTime = currentEventTime - lastEventTime
            vx = dx * 1000 / dTime
            vy = dy * 1000 / dTime
            lastX = newX
            lastY = newY
            lastEventTime = currentEventTime
        }

        val currImgX = this.x
        val currImgY = this.y

        val diffX = newX - currentX
        val diffY = newY - currentY
        val newImgX = currImgX + diffX
        val newImgY = currImgY + diffY

        this.x = newImgX
        this.y = newImgY

        yImageSlider.hideLeft.x += diffX
        yImageSlider.hideRight.x += diffX

        currentX = event.rawX
        currentY = event.rawY
    }

    override fun setImageBitmap(bm: Bitmap?) {
        super.setImageBitmap(bm)
        bitmapW = bm!!.width
        bitmapH = bm.height
    }



}
